Add a section for custom views and begin with `ToolBarView`'s documentation. (#19)
lite-xl/lite-xl.github.io: docs/developer-guide/views-examples.md

---
description: Learn about the most used custom view examples of Lite XL.
---

> ## Toolbar

Reviewer: Avoid creating consecutive heading like this. I'd prefer if you just discard the first heading and just put `Extending Views: ToolbarView`.

Author: What about something like the following:
![image](https://github.com/takase1121/lite-xl-docs/assets/80917080/e159432b-a94d-4b81-8182-cb6cb053bb88)
Because if I remove the `h2` heading, then the treeview to the right won't show the view name.

Reviewer: In this case, I think we can stil leave out h2. The right TOC goes to the tutorial directly (and I think Tutorial heading should be renamed to something else, like "Creating a plugin"

### Tutorial

> 

Reviewer: It's worth adding a screenshot of what ToolbarView looks like, because its frequently used but the name is obscure.

```lua
-- mod-version:3 -- lite-xl 2.1
```

> This is required to specify Lite XL version compatibility.

Reviewer: Once again, I think this doesn't have to be explained because it might be explained in a previous tutorial.

> The instructions that begin with `require` add necessary libraries.

Reviewer: The code block above didn't mention any `require`, so this is pointless here. I assumed that anyone reading would know a bit of lua. It might be worth pointing to the learning lua page.

```lua
local toolbar = common.merge({
  -- plugin configuration options go here

> }, config.plugins.build)

Reviewer: why is it merging `config.plugins.build`? Can you give a more substantial example, like a config option and optionally illustrate the settings GUI API here?

Author: I think the settings GUI API should be documented in a separate document, because It could be used by any custom View.
Also, the reference to `build` was an error on my part.

```

> The code above contains the (optional) configuration options of the plugin.

Reviewer: Optional?

The `get_plugin_directory()` function gets the plugin folder path and tells the 
plugin where to find the icon font.
The iteration code is used to check if every folder detected by `get_plugin_directory()` 
actually exists and if the folder doesn't exist, a `nil` value is returned.

`local ToolBar = ToolbarView:extend()` is used to create a new custom View and make it 

> inherit ToolbarView.

Reviewer: This is so confusing, the explanation with specific function names should go _after_ the code.

Explanation before the codeblock should explain roughly what the code does (often one to two sentences) and leave the detailed explanation after the codeblock, after the reader has read the code.

```lua
function ToolBar:new()
  ToolBar.super.new(self)
  self.toolbar_font = renderer.font.load(get_plugin_directory() .. PATHSEP .. "toolbar.ttf", style.icon_big_font:get_size())
  self.toolbar_commands = {

>     {symbol = "", command = "core:open-log"},

Reviewer: Can you make these symbol visible, for example with escape codes? I think this isn't needed in Lua, but it might be confusing if it just rendered as blank spaces.

Author: I don't understand what the problem is here.
`poetry` doesn't have access to my custom icon font, so the icon glyphs are always going to get displayed like that in the docs.

Reviewer: Right now the symbol string shows a box with a question mark. I'm not sure if it's possible to put a unicode literal like `\u1FFF` (pretty sure there is a function that converts a number to a unicode codepoint) instead of the string itself.

![image](https://github.com/takase1121/lite-xl-docs/assets/20792268/d692bc63-ab60-4c1d-acce-5232a1531e2c)

Author: I replaced the symbols with letters (`A`, `B`, ...), like how the default Lite XL toolbar does.

    {symbol = "", command = "core:new-doc"},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
  }
end
```

The code above creates an instance of our custom Toolbar view, inherits the ToolbarView super-constructor, 
specifies the desired icon font and assigns the icons to their respective commands.

`toolbar.example_toolbar_view = ToolBar()` saves the toolbar in an exported table, to allow any plugin to require it, thus accessing the bar.

```lua
toolbar.example_toolbar_node = TreeView.node.b:split("up", toolbar.example_toolbar_view, {y = true})
```

> This instruction splits the `TreeView` according to the `up` direction, adds the `toolbar.example_toolbar_view` and fixes its size along the y axis.

Reviewer: I think it would be better to refer to them as code instead of instructions. also `Y-axis` instead of `y axis`.

> `return toolbar` returns the toolbar table to Lite XL.

Reviewer: no need to explain this, also it doesn't return to Lite XL, it just returns to whatever the caller happens to be.

> To prepare a simple icon font with Fontello, you can do the following:

Reviewer: good place to use a heading here.

1. Go to [Fontello](https://fontello.com/)
2. Drag and drop the desired icons in the light-gray rectangle in the middle of the page, or 
upload your own svg files

> 3. Use `mouse-left-click` to select the icons you want to include in the font

Reviewer: IIRC that is not the correct format to use mouse click icons, it's `++left-button++`. I don't think it's necessary to tell users _which_ mouse button to click to select the icons.

4. Press on the `Customize Codes` tab on the top-left of the page and copy each icon glyph and 
paste it into the corresponding place in the `toolbar_commands` table
5. Specify the icon font name in the white text input field in the top right

> 6. Press the big red button in the top right corner of the page to create and download the icon font

Reviewer: Better give a name in case fontello adds two big red button on the top right center.

This section should have at least one screenshot to give the user some clues.

      return v
    end
  end
  return nil
end
```

### Creating the view

The code below creates an instance of our custom Toolbar View, inherits the ToolbarView super-constructor, 
specifies the desired icon font and assigns the icons to their respective commands.

```lua
local ToolBar = ToolbarView:extend()

function ToolBar:new()
  ToolBar.super.new(self)
  self.toolbar_font = renderer.font.load(get_plugin_directory() .. PATHSEP .. "toolbar.ttf", style.icon_big_font:get_size())
  self.toolbar_commands = {
    {symbol = "", command = "core:open-log"},
    {symbol = "", command = "core:new-doc"},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
  }
end
```

`local ToolBar = ToolbarView:extend()` is used to create a new custom View and make it 
inherit ToolbarView.

The `renderer.font.load` function assigns the icon font to the View.

The `get_plugin_directory()` function gets the plugin folder path.

### Adding the view

```lua
toolbar.example_toolbar_view = ToolBar()
toolbar.example_toolbar_node = TreeView.node.b:split("up", toolbar.example_toolbar_view, {y = true})
return toolbar
```

> The first line of code saves the toolbar in an exported table.

Reviewer: You don't have to explain the code line by line. Here you can specifically explain what the second line does.

For the explanation of the first and third, you can move it above the codeblock

---

> description: Learn about the most used custom view examples of Lite XL.

Reviewer: ```suggestion
description: Learn to extend a View in Lite XL for extra functionality.
```

---

## Example: ToolbarView

> brief introduction

Reviewer: Surely this is a placeholder for something right? You could describe what the plugin does.

    if system.get_file_info(v) then
      return v
    end
  end
  return nil
end
```

### Creating the view

The code below creates an instance of our custom Toolbar View, inherits the ToolbarView super-constructor, 
specifies the desired icon font and assigns the icons to their respective commands.

```lua
local ToolBar = ToolbarView:extend()

function ToolBar:new()
  ToolBar.super.new(self)
  self.toolbar_font = renderer.font.load(get_plugin_directory() .. PATHSEP .. "toolbar.ttf", style.icon_big_font:get_size())
  self.toolbar_commands = {
    {symbol = "", command = "core:open-log"},
    {symbol = "", command = "core:new-doc"},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
    {symbol = "", command = ""},
  }
end
```

`local ToolBar = ToolbarView:extend()` is used to create a new custom View and make it 
inherit ToolbarView.

The `renderer.font.load` function assigns the icon font to the View.

The `get_plugin_directory()` function gets the plugin folder path.

### Adding the view

The code below saves the toolbar in an exported table, to allow any plugin to require it, thus accessing the bar.

```lua

> toolbar.example_toolbar_view = ToolBar()

Reviewer: This codeblock can be combined with the next codeblock.

```

The code below splits the `TreeView` according to the `up` direction, adds the `toolbar.example_toolbar_view` and fixes its size along the Y-axis.

```lua
toolbar.example_toolbar_node = TreeView.node.b:split("up", toolbar.example_toolbar_view, {y = true})
```

> `return toolbar` returns the toolbar table to whatever function called it.

Reviewer: It might be worth mentioning that this would allow someone else to extend ToolbarView by `require "plugins.blabla"`

### Creating a custom icon font

To prepare a simple icon font with Fontello, you can do the following:

> 1. Go to [Fontello](https://fontello.com/)

Reviewer: Do not use this type of link.

Author: Right, sorry, I missed that one.
No inline links.

2. Drag and drop the desired icons in the light-gray rectangle in the middle of the page: ![Drag & Drop][drag-n-drop]

3. Select the icons you want to include in the font
4. Press on the `Customize Codes` tab on the top-left of the page and copy each icon glyph and 
paste it into the corresponding place in the `toolbar_commands` table

>     ![Customize Codes][customize-codes]

Reviewer: This screenshot is marginally better than nothing. Please put a screenshot of the entire web page and highlight the tab with a red box or something.

5. Specify the icon font name in the white text input field in the top right
6. Press the rectangular red button with `Download webfont` in the top right corner of the page to create and download the icon font: 

>     ![Red Button][red-button]

Reviewer: Refer to the previous comment.
